kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/collections/bounds.kt

```diff
@@ -5,52 +5,253 @@ import io.kotest.matchers.Matcher
+// "shouldContain true / shouldContain false"-type assertions
+
+/**
+ * Verifies that all elements are greater than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun ByteArray.shouldHaveLowerBound(value: Byte): ByteArray {
+   asList() should haveLowerBound(value, "ByteArray")
    return this
 }
 
-infix fun <T : Comparable<T>> Array<T>.shouldHaveLowerBound(t: T): Array<T> {
-   asList().shouldHaveLowerBound(t)
+/**
+ * Verifies that all elements are greater than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun ShortArray.shouldHaveLowerBound(value: Short): ShortArray {
+   asList() should haveLowerBound(value, "ShortArray")
    return this
 }
 
-infix fun <T : Comparable<T>, C : Collection<T>> C.shouldHaveLowerBound(t: T): C {
-   this should haveLowerBound(t)
+/**
+ * Verifies that all elements are greater than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun CharArray.shouldHaveLowerBound(value: Char): CharArray {
+   asList() should haveLowerBound(value, "CharArray")
    return this
 }
 
-fun <T : Comparable<T>, C : Collection<T>> haveLowerBound(t: T) = object : Matcher<C> {
-   override fun test(value: C): MatcherResult {
+/**
+ * Verifies that all elements are greater than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun IntArray.shouldHaveLowerBound(value: Int): IntArray {
+   asList() should haveLowerBound(value, "IntArray")
+   return this
+}
+
+/**
+ * Verifies that all elements are greater than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun LongArray.shouldHaveLowerBound(value: Long): LongArray {
+   asList() should haveLowerBound(value, "LongArray")
+   return this
+}
+
+/**
+ * Verifies that all elements are greater than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun FloatArray.shouldHaveLowerBound(value: Float): FloatArray {
+   asList() should haveLowerBound(value, "FloatArray")
+   return this
+}
+
+/**
+ * Verifies that all elements are greater than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun DoubleArray.shouldHaveLowerBound(value: Double): DoubleArray {
+   asList() should haveLowerBound(value, "DoubleArray")
+   return this
+}
+
+/**
+ * Verifies that all elements are greater than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun <T : Comparable<T>> Array<T>.shouldHaveLowerBound(value: T): Array<T> {
+   asList() should haveLowerBound(value, "Array")
+   return this
+}
+
+/**
+ * Verifies that all elements are greater than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun <T : Comparable<T>, C : Collection<T>> C.shouldHaveLowerBound(value: T): C {
+   this should haveLowerBound(value, null)
+   return this
+}
+
+/**
+ * Verifies that all elements are greater than or equal to [value].
+ *
+ * Passes if `this` is empty.
+ */
+infix fun <T : Comparable<T>, I : Iterable<T>> I.shouldHaveLowerBound(value: T): I {
```

Comment:
Do we actually need to define `I` here?

Reply:
Someone requested the underlying type to be returned due to having custom Iterable implementations. That way he could preserve the underlying type through assertion chain.

See https://github.com/kotest/kotest/issues/3960

Reply:
As returning the underlying type is the currently accepted behavior - and was so before this PR - I'll close this as resolved. Please reopen if you think there should be more discussion or if you think this should be changed to return Iterable again. But keep in mind that someone requested this behavior and it was accepted.